stor/dx.py

```diff
@@ -0,0 +1,513 @@
+        """Checks existence of the path.
+
+        Returns True if the path exists, False otherwise.
+
+        Returns:
+            bool: True if the path exists, False otherwise.
+        """
+        try:
+            # first see if there is a specific corresponding object
+            self.stat()
+            return True
+        except (NotFoundError, ValueError):
+            pass
+        # otherwise we could be a directory, so try to grab first
+        # file/subfolder
+        if self._is_folder():
+            try:
+                self.list(limit=1)
+                return True
+            except NotFoundError:
+                return False
+        return False
+
+    def stat(self):
+        if not self.__stat:
+
+            if self.canonical_resource:
+                self.__stat = dxpy.DXFile(dxid=self.canonical_resource,
+                                          project=self.canonical_project).describe()
+            else:
+                self.__stat = dxpy.DXProject(dxid=self.canonical_project).describe()
+        return self.__stat
+
+
+class DXVirtualPath(DXPath):
+    """Class Handler for DXPath of form 'dx://project-{ID}:/a/b/c' or 'dx://a/b/c'"""
+
+    @property
+    def virtual_project(self):
+        if utils.is_valid_dxid(self.project, 'project'):
+            return dxpy.DXProject(dxid=self.project).name
+        return self.project
+
+    @property
+    def virtual_resource(self):
+        return self.resource
+
+    @property
+    def virtual_path(self):
+        return self
+
+    @cached_property
+    def canonical_project(self):
+        """Returns the unique project that matches the name that user has view access to.
+        If no match is found, returns None
+
+        Raises:
+            DuplicateProjectError - if project name is not unique on DX platform
+            NotFoundError - If project name doesn't exist on DNAnexus
+        """
+        if utils.is_valid_dxid(self.project, 'project'):
+            return self.project
+        else:
+            try:
+                proj_dict = dxpy.find_one_project(
+                    name=self.project, level='VIEW', zero_ok=True, more_ok=False)
+                if proj_dict is None:
+                    raise ProjectNotFoundError('No projects were found with given name ({})'
+                                               .format(self.project))
+                return proj_dict['id']
+            except DXError as e:
+                raise DuplicateProjectError('Duplicate projects were found with given name ({})'
+                                            .format(self.project), e)
+
+    @cached_property
+    def canonical_resource(self):
+        """Returns the dx file-ID of the uniquely matched filename
+
+        Raises:
+            DuplicateError: if filename is not unique
+            NotFoundError: if resource is not found on DX platform
+        """
+        if not self.resource:
+            return None
+        if self._is_folder():
```

Comment:
when I look here, it feels like this should just raise a NotFoundError that specifies no data object could be found for the given path.

Reply:
Agreed and changed.